traits/trait_types.py

```diff
@@ -1953,19 +1961,51 @@ def __init__(self, *args, **metadata):
                     "when using the 'values' keyword"
                 )
         else:
-            default_value = args[0]
-            if (len(args) == 1) and isinstance(default_value, EnumTypes):
-                args = default_value
-                default_value = enum_default(args)
-            elif (len(args) == 2) and isinstance(args[1], EnumTypes):
-                args = args[1]
+            if len(args) < 1:
+                raise TraitError("Enum trait requires at "
+                                 "least 1 argument.")
+
+            elif len(args) == 1:
+                arg = args[0]
+                if isinstance(arg, EnumTypes):
+                    default_value = next(iter(arg), None)
+                    self.values = tuple(arg)
+
+                # Treat str, bytes and bytearray as discrete units,
+                # and not as a collection.
+                elif isinstance(arg, (str, bytes, bytearray)):
+                    default_value = arg
+                    self.values = {arg}
+
+                # Handle a collection
+                else:
+                    default_value = next(iter(arg), None)
+                    self.values = arg
+
+            elif len(args) == 2:
+                # If 2 args, the first is default, second is allowed values.
+                default_value = args[0]
+                allowed_vals = args[1]
+
+                # Treat str, bytes and bytearray as discrete units,
+                # and not as a collection.
+                if isinstance(allowed_vals, (str, bytes, bytearray)):
+                    self.values = tuple(args)
+
+                elif is_collection(allowed_vals):
+                    self.values = allowed_vals
```

Comment:
We should probably convert to a tuple here, to match the docstring (which says that `self.values` is a tuple).